### PR TITLE
Feature.cleanup

### LINKED ISSUE #66: Network configuration objects not removed when deleting load balancer
Network objects -- SNATs, tunnels, etc. -- are not removed when load balancer is deleted.

### PR DESCRIPTION
**Pull request by jlongstaf**
@richbrowne 
#### What issues does this address?
Fixes #66

#### What's this change do?
Cleans up subnet objects when members and loadbalancer are deleted.

#### Where should the reviewer start?
Start with lbaas_builder.py and look at how subnethints are constructed. Also,
closely review network_helper.py changes.

### CODE REVIEW

File: f5_openstack_agent/lbaasv2/drivers/bigip/snats.py [hunk 1]
```diff
@@ -266,6 +270,8 @@ def _delete_bigip_snats(self, bigip, subnetinfo, tenant_id):
             if not snatpool.members:
                 LOG.debug('Snat pool is empty - delete snatpool')
                 snatpool.delete()
+            else:
```
Comment (richbrowne) [inline f5_openstack_agent/lbaasv2/drivers/bigip/snats.py]: Exception handling?

File: f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py [hunk 2]
```diff
@@ -420,14 +443,23 @@ def arp_delete_by_subnet(self, bigip, subnet=None, mask=None,
 
         return self._delete_by_network(bigip, partition, network)
 
+    @log_helpers.log_method_call
     def _delete_by_network(self, bigip, partition, network):
```
Comment (richbrowne) [inline f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py]: Use name for function that is more meaningful.

File: f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py [hunk 3]
```diff
@@ -420,14 +443,23 @@ def arp_delete_by_subnet(self, bigip, subnet=None, mask=None,
...
+            arps = ac.get_collection(requests_params=params)
+        except HTTPError as err:
+            LOG.error("Error getting ARPs."
+                      "Repsponse status code: %s. Response "
+                      "message: %s." % (err.response.status_code,
+                                        err.message))
+            return mac_addresses
+
         for arp in arps:
             ad_rd_div = arp.ipAddress.find('%')
```
Comment (richbrowne) [inline f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py]: ad_rd_div???????

File: f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py [hunk 4]
```diff
@@ -468,11 +510,11 @@ def get_virtual_service_insertion(
...
 
         for virtual_server in virtual_services:
-            service = {'name': {}}
+            name = virtual_server.name
+            service = {name: {}}
             dest = os.path.basename(virtual_server.destination)
             (vip_addr, vip_port) = self.split_addr_port(dest)
 
-            name = self.strip_folder_and_prefix(virtual_server.name)
             service[name]['address'] = vip_addr
```
Comment (richbrowne) [inline f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py]: service is a little confusing without contxt.

File: f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py [hunk 5]
```diff
@@ -514,25 +559,38 @@ def add_fdb_entry(
...
-                                  % e.message)
-                        return False
-            return True
+        try:
+            tunnel = bigip.net.fdbs.tunnels.tunnel
+            if tunnel.exists(name=tunnel_name, partition=partition):
+                tunnel.load(name=tunnel_name, partition=partition)
+                tunnel.update(records=records)
+                if const.FDB_POPULATE_STATIC_ARP:
+                    if arp_ip_address:
```
Comment (richbrowne) [inline f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py]: This is always None?  (i.e. arp_ip_address)  This i guess is the ip address of the member.

File: f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py [hunk 6]
```diff
@@ -548,7 +606,7 @@ def delete_fdb_entry(
                                 partition=partition)
 
         records = self.get_fdb_entry(
-            bigip, tunnel_name, mac_address, partition=partition)
```
Comment (richbrowne) [inline f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py]: Generally, need to look for places where there is a named param in the function def and we are not explicitly citing that in the function call.

File: f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py [hunk 7]
```diff
@@ -559,11 +617,19 @@ def delete_fdb_entry(
             if len(records) == 0:
                 records = None
 
-        tunnel = bigip.net.fdbs.tunnels.tunnel
-        if tunnel.exists(name=tunnel_name, partition=partition):
-            tunnel.load(name=tunnel_name, partition=partition)
-            tunnel.update(record=records)
+        try:
+            tunnel = bigip.net.fdbs.tunnels.tunnel
```
Comment (richbrowne) [inline f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py]: Generally, need to look for resource acquisition code where we do not perform in try-catch.

File: f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py [hunk 8]
```diff
@@ -638,64 +705,96 @@ def delete_fdb_entries(self, bigip, tunnel_name=None, fdb_entries=None):
...
             partition=const.DEFAULT_PARTITION):
         """Delete all fdb entries."""
-        t = bigip.net.fdbs.tunnels.tunnel
-        t.load(name=tunnel_name, partition=partition)
-        LOG.debug(t.raw)
-        t.update(records=None)
+        try:
+            t = bigip.net.fdbs.tunnels.tunnel
+            t.load(name=tunnel_name, partition=partition)
+            LOG.debug(t.raw)
```
Comment (richbrowne) [inline f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py]: Do wee need to keep this around?

File: f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py (hunk 5 above)
Comment (jlongstaf) [inline f5_openstack_agent/lbaasv2/drivers/bigip/network_helper.py]: Not always None. This is the member address.

**Pull request merged**